src/sock.c

```diff
@@ -310,6 +314,88 @@ static int listen_on_one_socket(struct addrinfo *ad)
         return listenfd;
 }
 
+#ifdef SO_BINDTODEVICE
+static int listen_on_interface(const char *interface, uint16_t port)
+{
+        int listenfd;
+        int ret;
+        int ipv6=1;
+        const int on = 1;
+        const int off = 0;
+        struct sockaddr_storage addr;
+        struct ifreq ifr;
+
+        listenfd = socket(PF_INET6, SOCK_STREAM, 0);
+        if (listenfd == -1) {
+                listenfd = socket(PF_INET, SOCK_STREAM, 0);
+                if (listenfd == -1) {
+                        log_message(LOG_ERR, "socket() failed: %s", strerror(errno));
+                        return -1;
+                }
+                ipv6=0;
+        }
+
+        ret = setsockopt(listenfd, SOL_SOCKET, SO_REUSEADDR, &on, sizeof(on));
+        if (ret != 0) {
+                log_message(LOG_ERR,
+                            "setsockopt failed to set SO_REUSEADDR: %s",
+                            strerror(errno));
+                close(listenfd);
+                return -1;
+        }
+
+        strncpy(ifr.ifr_name, interface, sizeof(ifr.ifr_name));
+        if (setsockopt(listenfd, SOL_SOCKET, SO_BINDTODEVICE, &ifr, sizeof(ifr)) < 0) {
+               log_message(LOG_ERR, "setsockopt(SO_BINDTODEVICE,%s) failed: %s", interface, strerror (errno));
```

Comment:
I have not read the surrounding tinyproxy code, but I think it should be
```suggestion
               log_message(LOG_ERR, "setsockopt(SO_BINDTODEVICE,%s) failed: %s", interface, strerror(errno));
```